modules/ding_innskraning/ding_innskraning.module

```diff
@@ -0,0 +1,780 @@
+  $key = _ding_innskraning_get_key();
+  if (!$key) {
+    drupal_set_message(t('No key provided, authentication can not be finished.'), 'info');
+    drupal_goto('user/login');
+  }
+
+  $objKey = new XMLSecurityKey($signAlgorithm, array('type' => 'private'));
+  $objKey->loadKey($key, FALSE);
+
+  $msg = 'SAMLRequest=' . urlencode($samlRequest);
+  $msg .= '&RelayState=' . urlencode($relayState);
+  $msg .= '&SigAlg=' . urlencode($signAlgorithm);
+  $signature = $objKey->signData($msg);
+  return base64_encode($signature);
+}
+
+/**
+ * Get attribute by name from XML document.
+ */
+function _ding_innskraning_get_attribute_by_name($dom, $name) {
+  // Extract value with xpath.
+  $assertion = $dom->getElementsByTagName('Assertion')->item(0);
+  $attributeStatement = $assertion->getElementsByTagName('AttributeStatement')->item(0);
+  $attributes = $attributeStatement->getElementsByTagName('Attribute');
+  foreach ($attributes as $attribute) {
+    $attributePropName = $attribute->getAttribute('Name');
+    if ($attributePropName == $name) {
+      $attributeValue = $attribute->getElementsByTagName('AttributeValue')->item(0);
+      $value = $attributeValue->nodeValue;
+      return $value;
+    }
+  }
+}
+
+/**
+ * Get user agent.
+ */
+function _ding_innskraning_get_user_agent($dom) {
+  return _ding_innskraning_get_attribute_by_name($dom, 'UserAgent');
+}
+
+/**
+ * Try to login user by SSN number.
+ */
+function _ding_innskraning_activate_user_by_ssn($ssn) {
+  $auth_res = array(
+    'success' => TRUE,
+    'creds' => array(
+      'name' => $ssn,
+      // Random password. Shouldn't be used, but we'll make it un-guessable
+      // for good measure, in case someone messes up.
+      'pass' => md5(drupal_random_bytes(32) . 'no_password_for_the_wicked'),
+      // Note that user was logged in externally.
+      'external-auth' => TRUE,
+    ),
+    'user' => array(
+      'mail' => '',
+      'blocked' => FALSE,
+      'data' => array(
+        'display_name' => '',
+      ),
+    ),
+  );
+
+  /*
+   * We call two private ding_user functions to do the heavy lifting. Sadly,
+   * nobody has re-factored ding_user to have a simple login function, so we
+   * have to duplicate its work here.
+   */
+
+  // Create new account in Drupal and if one exists update it.
+  $account = _ding_user_create_account($ssn, $auth_res);
+
+  // Check that a profile exists for the user and if not create one.
+  _ding_user_create_profile2($account);
+
+  $form_state['uid'] = $account->uid;
+  user_login_submit(array(), $form_state);
+  return TRUE;
+}
+
+/**
+ * Verify the signatures and data from island.is.
+ */
+function _ding_innskraning_verify($xmlDoc) {
+  global $base_url;
+
+  // Load library.
+  libraries_load('xmlseclibs');
+
+  // Check SAML version.
+  if ($xmlDoc->documentElement->getAttribute('Version') != '2.0') {
+    throw new Exception('Unsupported SAML version');
```

Comment:
I am not a fan of throwing the generic exception but that seems to be the general pattern here 😕 .

Reply:
Agreed.